[TTS] GAN-based spectrogram enhancer

racoiaws opened the pull request:
# What does this PR do ?

Add a GAN-based spectrogram enhancer to NeMo.

The intended purpose is to add details to blurry spectrograms from TTS models like FastPitch.

**Collection**: TTS

# Changelog 
- Add the spectrogram enhancer NeMo model
- Add example config and training script

# Usage
Given a trained model,
```python
model = SpectrogramEnhancer.restore_from("checkpoint.nemo")
with torch.no_grad():
    # spectrograms :: B x 80 x L
    # lengths  :: B - the usual LongTensor with lengths or every spectrogram in the batch
    enhanced_spectrograms = model.forward(condition=spectrograms, lengths=lengths)
    # enhanced_spectrograms = B x 80 x L
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.


## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Based on https://github.com/lucidrains/stylegan2-pytorch (MIT License)

racoiaws commented:
Example: FastPitch trained on LibriTTS train-clean-100 + enhancer)
![1475274b-8f1b-4c4d-a1eb-53a78b14f32a](https://user-images.githubusercontent.com/8864149/206228168-2ab010a2-edbe-4c7c-a394-633957df6585.png)

subhankar-ghosh commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,382 @@
+
+import math
+from random import random, randrange
+from typing import Dict, List, Optional, TypedDict, Union
```
Remove unused imports.

subhankar-ghosh commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,382 @@
+            )
+
+        attn_prior, durs, speaker = None, None, None
+        (audio, audio_lens, text, text_lens, attn_prior, pitch, _, speaker,) = batch
```
Use this function instead for uniformity: https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/helpers/helpers.py#L564

subhankar-ghosh commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,382 @@
+
+        return self.move_to_correct_device((target, condition, lengths))
+
+    def training_step(self, batch, batch_idx, optimizer_idx):
```
You should add a corresponding `validation_step` as well.

racoiaws replied:
I'm not convinced validation step is necessary here - all of the losses are rather uninformative in deciding which model is "better"

rlangman replied:
Losses in GAN training are always difficult to interpret, but we should track the losses in wandb if only for debugging and making sure things do not become NaN and such.

racoiaws replied:
Will do after implementing preprocessing

racoiaws replied:
Current decision is not to

subhankar-ghosh commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,382 @@
+            n_bands=cfg["n_bands"], network_capacity=cfg["network_capacity"], channels=1, fmap_max=cfg["fmap_max"],
+        )
+
+    def init_spectrogram_model(self):
```
Do we really need a `spectrogram_model`, can this model not get spectrograms directly from the dataloader? Look at Vocoder dataloader. Or if you want to use TTSDataset class then I think you can use "log_mel". The spectrograms can be precomputed, this should remove dependency on any particular spectrogram generator.

racoiaws replied:
This is a major part of the training process, we can't do without a spectrogram model
![sg2-enhancer-training-150](https://user-images.githubusercontent.com/8864149/207025284-efd9cf2f-c2c1-421f-b2d6-05eec84d06c7.png)

subhankar-ghosh replied:
Well, essentially what you are doing in this model is mel to mel. The goal is to not have any dependency on the spectrogram generator. Here FastPitch is frozen and is not trained as part of the spectrogram enhancer.

In HiFiGAN training also the input is mel, which are precomputed, which can be done here as well. 
Before training, do inference on FastPitch or Any spectrogram generator model to precompute the mels. Now, these mels can be an input to the spectrogram enhancer. In your PR as well it is part of the `prepare_batch` method, which means it is not part of training but used in preparing data, which in turn means this can be precomputed. See the https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/torch/data.py#L995 part of the VocoderDataset class that loads precomputed spectrograms.

During training your batch can look like this: `(precomputed_spec, groundtruth audio, len, ...)`
You can use `LogMel` ttsdatatype to load your mels and incorporate the padding logic - https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/torch/data.py#L514

racoiaws replied:
No. Precomputing spectrograms is very inconvenient (and also requires lots of space). I speak from experience.

Current implementation is the most clean and convenient way of doing this from user point of view.

racoiaws replied:
The main goal was to completely repeat data setup from FastPitch training: to use the same texts, audio and ground-truth precomputed features like pitch (and I guess energy, with new revisions). This gets us would get natural-sythesize) spectrogram pairs and minimizes chances of the GAN learning to fix pitch or other features. 
I figured the most easy and natural way to implement this is just to run FastPitch on its underlying TTSDataset, which is the intended way to train this model.

The initial precomputing script I used just initialized the TTSDataset and iterated through it, saving results. Same thing happening here, but inside PTL.

I haven't measured overhead, but I also did not notice any significant slowdowns while doing experiments for our paper
compared to precomputed spectrograms. Slower training is solved by waiting a bit more, but lack of storage is solved only by buying more storage.
Same thing with GPU memory though... I'll look at memory usage and get back to you on this

Note that `generate_mels.py` mirrors some of TTSDataset internals but does not use it directly. Handling of sup data mangled names is a pain, so it would be a full rewrite or just a separate script.

subhankar-ghosh replied:
Also, one very important aspect of spectrogram enhancer is that it should be independent of spectrogram generator. Why do we want to tie to any particular spectrogram generator (FastPitch)? If we want it to work with FastPitch only then we should make it a feature of FastPitch class. 

Having precomputed mels will enable us to use it for any spectrogram generator model. I understand that the experiments were done using FastPitch as spec generator, but when we release it on NeMo we should make it generic (so that it can be used with any spec gen model) as much as possible in my opinion.

subhankar-ghosh replied:
> Note that generate_mels.py mirrors some of TTSDataset internals but does not use it directly. Handling of sup data mangled names is a pain, so it would be a full rewrite to TTSDataset (questionable) or just a separate script.

`generate_mels.py` is just doing inference to generate the mels. You don't really need any other supplementary data other than `attn_prior` and `speaker` both of which are supplied. Pitch, energy, duration etc are predicted by the model.

racoiaws replied:
> Why do we want to tie to any particular spectrogram generator (FastPitch)? 

We really do not, it's just an example we're discussing. To use any other spectrogram generator just implement the boilerplate for converting a collated batch from TTSDataset to batch of paired real-fake spectrograms and it's good to go. Unfortunately they have different inputs (and interfaces) and thus need to be supported separately.

> generate_mels.py is just doing inference to generate the mels. You don't really need any other supplementary data other than attn_prior and speaker both of which are supplied. Pitch, energy, duration etc are predicted by the model.

Yes, and that's why it is not suitable for training this model. The more ground-truth features (pitch, energy), the better; durations also need to match. We need resynthesis using a TTS model.

rlangman replied:
There should not be any functional difference from the approaches. I guess generate_mels.py is not using ground truth information, but it should have it as a configuration. When doing GAN based fine-tuning with the spectrogram (whether through StyleGan or indirectly HiFi-Gan fine-tune) it is not 100% clear to me which would be better. Using ground truth feature input is definitely safer but can limit the potential quality improvements.

So precomputing and making generate_mels.py more generic is one possible solution. It sounds like the other solution you are proposing is to turn "prepare_batch(self, batch)" into an abstract method or separate class that someone can implement with model-specific logic.

subhankar-ghosh commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,382 @@
+        self.init_spectrogram_model()
+
+        self.G = Generator(
+            n_bands=cfg["n_bands"],
```
try to maintain uniformity across all models by using `cfg.n_bands` instead of `cfg["n_bands"]`

racoiaws replied:
Done

subhankar-ghosh commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,382 @@
+            targets.append(mels[i, :, :l].T)
+            conditions.append(mels_pred[i, :, :l].T)
+
+        target = nn.utils.rnn.pad_sequence(targets, batch_first=True)
```
Best to keep these kind of padding and data manipulation logic in dataloader.

racoiaws replied:
I initially wrote it as a separate dataloader and gave up after some debugging. 
In this implementation data loading is delegated to underlying spectrogram generator. Which ended up much more robust and clean.

Should I move it to its own class? Still would be a regular class, not a torch/nemo dataset or dataloader or collate_fn

subhankar-ghosh replied:
https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/torch/data.py#L995
Take a look at this part, it can help with padding.

racoiaws replied:
Sorry, I don't understand your last message about padding, what is your suggestion or what is the issue.

racoiaws replied:
Thanks Ryan, both are good points 

I guess at the initial research stage I copied messy code from the precomputing spectrograms script and reassembled tensors just to be absolute sure about data format. These were B x 1 x D x L for 2d convs.

Will remove redundancies

racoiaws replied:
Done but will be replaced with a preprocessing step

subhankar-ghosh commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,382 @@
+    return fake.mean()
+
+
+def hinge_loss(real, fake):
```
loss should be part of the class. like `self.loss_fn = hinge_loss`
Also if this is simple hinge loss then why not use predefined hinge loss from torch? If it is your own version (not exactly hinge loss) then name it accordingly. It can be confusing to users.

racoiaws replied:
Did not find anything named "hinge loss" in the docs

This is copied from https://github.com/lucidrains/stylegan2-pytorch

racoiaws replied:
Moved to a class and made a field inside the model

subhankar-ghosh commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,382 @@
+    return (F.relu(1 + real) + F.relu(1 - fake)).mean()
+
+
+def consistency_loss(condition, output, lengths):
```
same as above. All losses should be part of the class.

racoiaws replied:
Done

subhankar-ghosh commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,382 @@
+            d_loss = hinge_loss(real=real_logits, fake=fake_logits)
+            self.log("d_loss", d_loss, prog_bar=True)
+
+            if batch_idx % 4 == 0:
```
What is the logic behind this if statement? Have a comment/note explaining this. Or have a method with the proper name. Like:
```
def if_<condition_name>(self, batch_idx):
    return batch_idx % 4 == 0

if self.if_<condition_name>(batch_idx):
   ...
```

racoiaws replied:
`"gradient_penalty_loss_every_n_steps"`

subhankar-ghosh commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,382 @@
+            enhanced_spectrograms = self.forward(condition, lengths, mixing=True, normalize=False)
+            fake_logits = self.D(enhanced_spectrograms, condition, lengths)
+            g_loss = gen_hinge_loss(fake_logits)
+            c_loss = 10 * consistency_loss(condition, enhanced_spectrograms, lengths)
```
reason behind 10? Hardcoding should be avoided. Make it a class variable with proper name.
loss should be part of class.

racoiaws replied:
Good catch, this is supposed to be a parameter

racoiaws replied:
Done

racoiaws replied:
`"consistency_loss_weight"`

subhankar-ghosh commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,382 @@
+    return (pl_grads ** 2).sum(dim=2).mean(dim=1).sqrt()
+
+
+def gen_hinge_loss(fake):
```
name of the method and what it is doing are not the same. can be confusing. You can get rid of the method and replace with just `variable.mean()`

subhankar-ghosh commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,382 @@
+            return g_loss + c_loss
+
+    def configure_optimizers(self):
+        generator_opt = torch.optim.Adam(self.G.parameters(), lr=self._cfg["lr"], betas=(0.5, 0.9),)
```
optimizers should be initialized from `cfg` and hardcoding needs to be avoided.

racoiaws replied:
Done

subhankar-ghosh commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,382 @@
+
+    @classmethod
+    def list_available_models(cls):
+        return None
```
return an empty list.

racoiaws replied:
Done

subhankar-ghosh commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,382 @@
+        self,
+        condition: Tensor,
+        lengths: Tensor,
+        zs: Optional[List[Tensor]] = None,
```
All parameters need dimension and explanation. Naming can be better - what is zs, ws, etc. needs to be explained.

subhankar-ghosh replied:
https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/models/fastpitch.py#L290
for dim=1

racoiaws replied:
No, I'm talking about a tensor of B x 1 x D x L - I did not find a way of expressing the 1 requirement in a NeuralType.

I guess result is good for the end user - no need to care about this extra dimension.

subhankar-ghosh commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,382 @@
+        spectrogram = spectrogram + self._cfg["spectrogram_min_value"]
+        return spectrogram
+
+    def generate_zs(self, batch_size: int = 1, mixing: bool = False):
```
naming can be better - zs?

subhankar-ghosh commented on nemo/collections/tts/modules/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,355 @@
+from math import log2
+from typing import List
+
+import torch
```
Needs to be taken care of.

subhankar-ghosh replied:
This means when you are importing `import torch` you should call all functionalities of torch with full name like use use `torch.Tensor` and `torch.nn` instead of `from torch import Tensor, nn`

racoiaws replied:
> you should call all functionalities of torch with full name

Is is a NeMo code style requirement? I see nothing wrong with it, my preference would be is "ignore / won't fix".

rlangman replied:
I am just thinking things up on the spot right now. I am not an expert with PyTorch conventions, but I have never seen such layout/structure/usage with any other library.

I would default to just importing those 2 and  providing full paths beyond that, for clarity. The usual alternative, as mentioned in the code style warning, is to import all methods `from torch import Tensor, nn, randn, ...` but that sounds like a bad idea because so many torch functions like randn are the exact same name as equivalent numpy methods.

I see no consistency on this in the rest of NeMo codebase. It probably does not matter.

racoiaws replied:
Done

Left only `import torch` and `import torch.nn.functional as F`

subhankar-ghosh commented on requirements/requirements_tts.txt:
```diff
@@ -1,6 +1,7 @@
 attrdict
 einops
 inflect
+kornia
```
we need tests to verify that this does not affect other dependencies and break any test cases.

racoiaws replied:
How would one do that? CI says tests pass

It is used for a single op though, could be rewritten in PyTorch

subhankar-ghosh commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,382 @@
+
+        self.init_spectrogram_model()
+
+        self.G = Generator(
```
G, D needs to be instantiated from the yaml config.

racoiaws replied:
You mean passing a DictConfig to them? What is the rationale? Neither of them are ModelPT

subhankar-ghosh replied:
Not DictConfig object but providing the configs of your module in the yaml file like , I mean users should have the option to change G, D from the yaml file instead of touching the code. In the future, if someone wants to use a different Generator or Discriminator, it should be easier for them to just implement it and change in the yaml file instead of importing the modules in this file and changing the instantiation here. Doesn't really matter if they are inheriting ModelPT or not.

Example: duration predictor is instantiated here https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/tts/models/fastpitch.py#L143
And the corresponding config is provided here https://github.com/NVIDIA/NeMo/blob/main/examples/tts/conf/fastpitch_align_44100.yaml#L198
Now, if I want to use a different module as duration predictor I do not need to change the code, all I will need to do is to change the yaml file. Makes it easy and clean.

racoiaws replied:
> I mean users should have the option to change G, D from the yaml file instead of touching the code.

> In the future, if someone wants to use a different Generator or Discriminator, it should be easier for them to just implement it and change in the yaml file instead of importing the modules in this file and changing the instantiation here.

We use a specific config of a very specific generative model (StyleGAN 2), modified for working with sequences. Using your own generator or discriminator is discouraged.

Personally to me, this YAML-oriented design caused a lot of frustration when I was starting with NeMo and FastPitch. Some issues off the top of my head:
1) Readability. There is no class name in code, just a bare `instantiate` and a config field. It instantiates _something_, and one has to jump between the config and the code. (Half a year passed, and I'm still doing this)
2) Explorabilily. Which modules exactly can I put there? The ones with the name "Predictor" in it (FP example)?
3) Extensibility is illusionary. Want your new replacement submodule to take some extra input? Modify Python code, YAML can't do this. Or write a Python monster to make this YAML-configurable.

Overall, the "just change it in the yaml file, easy and clean" feature actually comes at cost of spiraling cognitive load when trying to understand the code. It also blurs the border between YAML configuration and Python programming, making both harder to work with.

While there are strong benefits for models like FastPitch, I don't think it is warranted here at all.

racoiaws replied:
Not exposing G and D in the YAML config was a conscious decision from my side. Unless it's an absolute requirement and blocks merging, I'd rather not do it.

rlangman replied:
To not instantiate them with configurable parameters in the YAML would go against the NeMo design pattern. I do not know if it is the optimal approach, and its hard to say because in general we are clearly executing on it poorly (eg. 90% of the config files are copy-pasted making them way too large. NN layers which could be shared between models are instead copy-pasted between model definitions. The interfaces for the modules are not generic or documented).

But I think arguing that is out of the scope for a PR. Any individual PR should have to follow the coding styles and design patterns of the code base as decided by the organization.

racoiaws replied:
> would go against the NeMo design pattern

> But I think arguing that is out of the scope for a PR. Any individual PR should have to follow the coding styles and design patterns of the code base as decided by the organization.

OK, good points.

racoiaws replied:
Done

racoiaws commented on nemo/collections/tts/modules/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,355 @@
+        return x
+
+
+class Conv2DMod(nn.Module):
```
Modulated convolution - https://arxiv.org/abs/1912.04958

Should I rename it to Conv2DModulated?

subhankar-ghosh replied:
Sure, that sounds good. Also in the docstring add the link to the paper.

racoiaws replied:
Done

racoiaws commented on nemo/collections/tts/modules/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,355 @@
+
+        self.blocks = nn.ModuleList(blocks)
+
+        chan_last = filters[-1]
```
Done

racoiaws commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,382 @@
+
+        self.init_spectrogram_model()
+
+        self.G = Generator(
```
I'm on the fence on this - official and unofficial implementations of StyleGAN use G and D, so it reads easier for people familiar with that code

racoiaws replied:
It is ubiquitously G and D in the official NVIDIA implementations, e.g. https://github.com/NVlabs/stylegan3/blob/main/training/training_loop.py

racoiaws replied:
Renaming done

racoiaws commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,395 @@
+from random import random, randrange
+from typing import List, Optional
+
+import torch
```
I do not understand what is the issue here

racoiaws commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,395 @@
+
+# The following is largely based on code from https://github.com/lucidrains/stylegan2-pytorch
+
+import math
```
Will be fixed in next push

racoiaws replied:
Done

redoctopus commented on tests/collections/tts/test_spectrogram_enhancer.py:
```diff
@@ -0,0 +1,186 @@
+    return DictConfig(config)
+
+
+@pytest.fixture
+def fastpitch_model():
+    model = FastPitchModel.from_pretrained(model_name="tts_en_fastpitch")
+    return model
```
I wonder if this should be in a separate test file since this does just test FastPitch NGC checkpoint loading in general.

racoiaws replied:
This not a test, this is a fixture, a fixture is not supposed to test something by itself

Is designing reusable testing code something worth of effort? I've been taught e.g. copying-pasting tests is an OK thing to do

redoctopus commented on tests/collections/tts/test_spectrogram_enhancer.py:
```diff
@@ -0,0 +1,186 @@
+@pytest.mark.with_downloads
+@pytest.mark.unit
+def test_enhancer_with_fastpitch(enhancer_with_fastpitch: SpectrogramEnhancerModel):
+    # long setup for enhancer_with_fastpitch, around 1 min
```
Should this be a unit test if it takes 1m+ to set up and run?

racoiaws replied:
Good question! Do we have a working way to say THIS TAKES LONG in our testing setup? The `.unit` mark here is just copy-paste

racoiaws replied:
@rlangman please take a look at how tests are made, thanks

redoctopus replied:
Ah, I found this in the development docs:
> Unit tests must be fast: <1s ideally, <10s at worst, and >10s needs to have an exception from NeMo Core team.

So I guess this should be an integration test ("These tests are longer running and are used to confirm a behavior.").

redoctopus replied:
You should be able to mark them with `@pytest.mark.integration`, but I only see a few Hydra tests that use it in NeMo. @okuchaiev can you advise on best practice here? Should we be using this marker more, or is something else more appropriate?

racoiaws replied:
Removed the tests because the configuration they used was also removed

racoiaws commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,428 @@
+        """
+        Generator forward pass. Noise inputs will be generated.
+
+        condition: batch of blurred spectrograms
```
Since it came up in the review process, it's probably too abstract. I'll rename it to something more concrete, `blurred_spec` sounds good.

redoctopus replied:
Seems fine to me :)

racoiaws commented on examples/tts/conf/spectrogram-enhancer.yaml:
```diff
@@ -0,0 +1,67 @@
+  mixed_prob: 0.9
+  fmap_max: 192
+
+  consistency_loss_weight: 10.0  # less for clean datasets, higher for noisier
+  gradient_penalty_loss_weight: 10.0
```
For consistency loss, 1.0 to 20.0 is probably a good range (eyeballing). Will put it in the comment.

For gradient penalty loss, I honestly have no idea. This is the default value from the original repo and it works

racoiaws replied:
Comments done

racoiaws commented on examples/tts/conf/spectrogram-enhancer.yaml:
```diff
@@ -0,0 +1,67 @@
+  # Needed for treating spectrograms as images with pixel values around [0, 1].
+  # Given values are for LibriTTS, might still work for your data.
+  # In case of training not converging, calculate min and max values over your TTSDataset and update the config.
+  spectrogram_min_value: -13.18
+  spectrogram_max_value: 4.78
```
Will do

racoiaws replied:
Done

racoiaws commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,428 @@
+    def move_to_correct_device(self, e):
+        return type_as_recursive(e, next(iter(self.G.parameters())))
+
+    def normalize_spectrograms(self, spectrogram: Tensor) -> Tensor:
+        spectrogram = spectrogram - self._cfg.spectrogram_min_value
+        spectrogram = spectrogram / (self._cfg.spectrogram_max_value - self._cfg.spectrogram_min_value)
+        return spectrogram
+
+    def unnormalize_spectrograms(self, spectrogram: Tensor) -> Tensor:
+        spectrogram = spectrogram * (self._cfg.spectrogram_max_value - self._cfg.spectrogram_min_value)
+        spectrogram = spectrogram + self._cfg.spectrogram_min_value
+        return spectrogram
```
Personally not a fan of abstracting one-liners like this. Makes me jump around the project and lose context.

Will surely make the `norm(unnorm(x)) == x` and `unnorm(norm(x)) == x` tests

redoctopus replied:
Sure, that's reasonable. Thanks!

racoiaws replied:
Done

Also added masking, better user interface

racoiaws commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,428 @@
+        return self.weight * ((gradients.norm(2, dim=1) - 1) ** 2).mean()
+
+
+class GeneratorLoss(nn.Module):
+    def __call__(self, fake_logits):
+        return fake_logits.mean()
+
+
+class HingeLoss(nn.Module):
+    def __call__(self, real_logits, fake_logits):
+        return (F.relu(1 + real_logits) + F.relu(1 - fake_logits)).mean()
+
+
+class ConsistencyLoss(nn.Module):
```
Will do

racoiaws commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,428 @@
+from nemo.utils import logging
+
+
+def type_as_recursive(e, source: Tensor):
```
Will do

racoiaws commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,428 @@
+
+        mels_pred, *_ = self.spectrogram_model(
+            text=batch["text"],
+            durs=None,
+            pitch=batch["pitch"],
+            speaker=batch.get("speaker"),
+            pace=1.0,
+            spec=mels if self.spectrogram_model.learn_alignment else None,
```
We use the model to "resynthesize" input dataset, so the more ground-truth features the better.

`learn_alignment=True` means we FastPitch learns to compute alignment from input spectrogram and encoded text, which happens during training and exactly what we need here, because we need tts output spectrograms to have the same length as the input.

There is a bug here though: a model with `learn_alignment=False` is not handled correctly. `durs=None` always, so we most likely get exception somewhere down the line.

racoiaws commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,428 @@
+        Generator forward pass. Noise inputs will be generated.
+
+        condition: batch of blurred spectrograms
+        lenghts: length for every spectrogam in the batch
```
Fixed, thanks

racoiaws commented:
Hi folks @rlangman @subhankar-ghosh
Please take a look at data preprocessing & loading update.

- added a script for resynthesis 
- implemented a Dataset for loading pairs of real-resynthesized spectrograms
- removed the spectrogram generator inside the model, switched to the new Dataset
- updated example config and removed tests

Couple of commits got squashed into the last one unfortunately

racoiaws commented on scripts/dataset_processing/tts/resynthesize_dataset.py:
```diff
@@ -0,0 +1,244 @@
+from nemo.collections.tts.helpers.helpers import process_batch, to_device_recursive
+from nemo.collections.tts.models import FastPitchModel
+from nemo.collections.tts.models.base import SpectrogramGenerator
+from nemo.utils import logging
```
resolved

rlangman reviewed (commented):
Overall, looks good to me. Thanks for the effort refactoring the precompute and data loading. Just a few minor comments.

racoiaws commented on tests/collections/tts/test_spectrogram_enhancer.py:
```diff
@@ -0,0 +1,159 @@
+    path = tmp_path / "test-enhancer-save-load.nemo"
+
+    enhancer.save_to(path)
+    return SpectrogramEnhancerModel.restore_from(path)
```
I think this return value was used in another test that was deleted

racoiaws replied:
Removed

racoiaws commented on scripts/dataset_processing/tts/resynthesize_dataset.py:
```diff
@@ -0,0 +1,243 @@
+            yield self.resynthesize_batch(batch)
+
+
+def read_manifest(path: Path) -> Dict[str, str]:
```
Good catch, was not aware of that, will do

racoiaws commented on nemo/collections/tts/losses/spectrogram_enhancer_losses.py:
```diff
@@ -0,0 +1,102 @@
+
+    def __call__(self, condition, output, lengths):
+        *_, w, h = condition.shape
+        w, h = w // 4, h
```
Nothing, just visual coherence. Even more explicit statement we don't do anything with `h`, I guess.

racoiaws commented on scripts/dataset_processing/tts/resynthesize_dataset.py:
```diff
@@ -0,0 +1,243 @@
+            pred_mel = pred_mels[i, :, :length].clone().numpy()
+            true_mel = true_mels[i, :, :length].clone().numpy()
+
+            pred_mel_path = output_folder / "mels" / f"{filename}.mel.npy"
+            true_mel_path = output_folder / "mels" / f"{filename}.mel.gt.npy"
```
Will do

racoiaws replied:
It's already in a "mels" folder, so changed to just `.npy` and `_gt.npy`

racoiaws commented on scripts/dataset_processing/tts/resynthesize_dataset.py:
```diff
@@ -0,0 +1,243 @@
+
+    output_manifest = []
+    output_json_manifest = output_folder / f"{input_json_manifest.stem}_mel{input_json_manifest.suffix}"
+    (output_folder / "mels").mkdir(exist_ok=True, parents=True)
```
Will do

racoiaws commented on nemo/collections/tts/modules/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,372 @@
+        return x, rgb
+
+
+def mask(tensor, lengths):
```
Done

rlangman reviewed (commented):
Looks good to me. Will let @subhankar-ghosh review again before approving.

subhankar-ghosh commented on examples/tts/conf/spectrogram-enhancer.yaml:
```diff
@@ -0,0 +1,86 @@
+  train_ds:
+    dataset:
+      _target_: "nemo.collections.tts.torch.data.PairedRealFakeSpectrogramsDataset"
+      manifest_filepath: "/home/user/train-clean-100.manifest_mel.json"
```
Let's not put paths over here. Change this to a mandatory field i.e. ???

racoiaws replied:
Good catch, thank you

This is a spillover from my test docker environment, definitely supposed to be `???`

racoiaws replied:
Fixed

subhankar-ghosh commented on nemo/collections/tts/models/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,313 @@
+            dataset, collate_fn=dataset.collate_fn, **train_data_config.dataloader_params
+        )
+
+    def setup_validation_data(self, val_data_config):
```
We can go two ways here:
1) Remove `setup_validation_data` because the is no `val_ds` in the config, nor is there a validation step implementation.
OR
2) Add `val_ds` section to config as well as a validation_step to the model class.

racoiaws replied:
Let's say we compute discriminator loss on some held-out set. I am extremely skeptical whether it is a sensible way to choose between two models based on that loss. Until proven on a couple of downstream setups which use output of this GAN, of course.
I don't think validation steps are of value here, even though the generation is conditional. A validation step might be introduced in the future, if (when) we get some positive evidence, but for now let's keep it simple.

Replaced with a `pass` and a comment summing up above. This is an `@abstractmethod` in `ModelPT`, so there has to be _some_ implementation

racoiaws replied:
Basically went with option 1.

subhankar-ghosh commented on nemo/collections/tts/modules/spectrogram_enhancer.py:
```diff
@@ -0,0 +1,359 @@
+        batch_size, _, _, max_length = condition.shape
+
+        x = self.initial_block.expand(batch_size, -1, -1, -1)
+        x = x[:, :, :, : max_length // 16]
```
what is 16?

racoiaws replied:
This depends on the fact that we have four 2x upsample steps, so 2**4

Supposed to depend on number of blocks, also a good catch

racoiaws replied:
Done, made a computable property for this

subhankar-ghosh commented:
Looks much better now. Left a few comments/Nit picks.

subhankar-ghosh commented:
LGTM.

subhankar-ghosh reviewed (approved):
LGTM

subhankar-ghosh commented:
Future Work @racoiaws :
1) Let's create a tutorial on how to use it OR you can add a section to FastPitch Finetuning tutorial about the spectrogram enhancer as an alternative to fine-tuning hifigan.
2) Let's also show users the comparison of audio produced by spec enhancer + frozen vocoder and fine-tuned vocoder.

racoiaws commented:
Thanks everyone for your review @redoctopus @subhankar-ghosh @rlangman

> Future Work @racoiaws :
> 
> 1. Let's create a tutorial on how to use it OR you can add a section to FastPitch Finetuning tutorial about the spectrogram enhancer as an alternative to fine-tuning hifigan.
> 2. Let's also show users the comparison of audio produced by spec enhancer + frozen vocoder and fine-tuned vocoder.

OK.
